core/dbt/deps/tarball.py

```diff
@@ -0,0 +1,238 @@
+import os
+# import shutil
+import hashlib
+import tarfile
+from tempfile import NamedTemporaryFile
+from typing import Optional, Union
+from dbt.clients import system
+from dbt.deps.base import PinnedPackage, UnpinnedPackage, get_downloads_path
+from dbt.exceptions import (
+    DependencyException, 
+    package_structure_malformed, 
+    package_sha1_fail
+)
+from dbt.events.functions import fire_event
+from dbt.events.types import Sha1ChecksumPasses, UntarProjectRoot
+from dbt.config import Project
+from dbt.contracts.project import (
+    ProjectPackageMetadata,
+    TarballPackage,
+)
+
+
+TARFILE_MAX_SIZE = 1 * 1e+6  # limit tarfiles to 1mb
+
+
+class TarballPackageMixin:
+    def __init__(self, tarball: str) -> None:
+        super().__init__()
+        self.tarball = tarball
+
+    @property
+    def name(self):
+        return self.tarball
+
+    def source_type(self) -> str:
+        return 'tarball'
+
+
+class TarballPinnedPackage(TarballPackageMixin, PinnedPackage):
+    def __init__(
+        self,
+        tarball: Union[str, tarfile.TarFile],
+        sha1: Optional[str] = None,
+        subdirectory: Optional[str] = None,
+    ) -> None:
+        super().__init__(tarball)
+        self.sha1 = sha1
+        self.subdirectory = subdirectory
+        # self.tarfile_bin = self.get_tarfile()
+        self.get_tarfile()
+        ''' init tarfile in class, and use tarfile as cache.
+            This simplifies tempfile handling, and prevents multiple hits to
+              url, since we need the file for two seperate operations. One
+              for metadata fetch (in style of other packages) and once for
+              for install for install. '''
```

Comment:
I've removed caching the tarfile as part of the TarballPinnedPackage, and now the url is hit twice (once to get metadata about the package, and a second time to download the package). It's only a minor changes to code, but if you think that's the right tradeoff, then no problem for me. 

For the more complex topic of reusing registry or git package code - 
I've spent quite a lot of time trying to recycle the registry code, prior to coming to the current solution. However there are some fundamental differences in how this is handled vs registry. Registry has all the lovely json metadata that coordinates the tar download, and naming (as well as enforcing tar file structure, and not allowing subdirectories. Nothing to recycle from registry client, since it's all loading and parsing the json. And the registry install is super simple, since registry api/json handles any unknowns. File source structure and destination name are taken care of, and there is no concept of package subdirectory.
Git provides a good 'cousin' example, in that it allows for subdirectories, and has no metadata source like registry. I've recycled the subdirectory scanning from the git code. However git uses clone and copy for the download and install, so none of the install routine can be copied here. 

Here's a *rough* sketch comparing registry install vs tarball install. Let me know where you have any ideas on streamlining the code between the two. My approach so far has been built on the premise that the registry metadata creates a very simple solution, and so the registry install routine is very straight forward, but the registry install routine overfits to scenario of not needing to 'discover' any metadata. Tarball, on the other hand, given added unknowns from not having metadata, becomes unique in having to discover the metadata from the file structure, and then in turn use that to unpack the package in an intelligent way. Additionally we are doing tarfile verification, sha1 verification, and size verification, which isn't happening for any of the registry (or git, or local) routine, so new code having to be added there. 
![Untitled drawing](https://user-images.githubusercontent.com/105436/155035578-41d5afc8-acbc-4fb7-84cf-16bd64104e38.png)

Phew! ok lots there, but my tldr is please let me know where you see opportunity to streamline/recycle from registry install routine. I've given this a lot of thought already, and do not see any obvious patterns, aside from requiring a third style of install (tarball) that is more of a hybrid of of parts of registry and Git install, than overly similar to either.

Reply:
I've just finished a total rewrite with this `If this behaved the same as the registry package more of the code could be shared to have a single path to follow.` in mind. So ready for a new review :)